python/adelphi/bin/adelphi

```diff
@@ -103,104 +132,124 @@ def unique(iterable, key=None):
+
 
+def get_keyspaces(props, cluster):
+    keyspace_names = props["keyspace-names"]
+    metadata = cluster.metadata
+    keyspaces = build_keyspace_objects(keyspace_names, metadata)
 
-def validate_keyspaces(keyspace_objs):
-    if len(keyspace_objs) == 0:
-        log.info("No keyspaces selected")
+    if len(keyspaces) == 0:
+        log.info("No valid keyspaces selected")
         exit(1)
 
+    log.info("Processing the following keyspaces: %s", ','.join((ks.name for ks in keyspaces)))
+    return {anonymize_keyspace(ks) if props['anonymize'] else ks : hash_keyspace_name(ks) for ks in keyspaces}
+
 
 def get_cluster_metadata(cluster):
     hosts = cluster.metadata.all_hosts()
     unique_dcs = unique((host.datacenter for host in hosts))
-    return {"host_count": len(hosts), "dc_count": sum(1 for i in unique_dcs)}
-
+    return {"host-count": len(hosts), "dc-count": sum(1 for i in unique_dcs)}
 
-def gather_keyspaces(ctx):
-    partial_fn = partial(get_keyspaces, ctx)
-    keyspaces = with_cluster(partial_fn, ctx)
 
-    log.info("Exporting schema for keyspaces %s", ','.join((k.name for k in keyspaces)))
+def build_keyspace_map(props):
+    partial_fn = partial(get_keyspaces, props)
+    return with_cluster(partial_fn, props)
 
-    validate_keyspaces(keyspaces)
-    return keyspaces
 
-
-def gather_metadata(ctx):
-    metadata = {k : ctx.obj[k] for k in ["purpose", "maturity"]}
-    metadata.update(with_cluster(get_cluster_metadata, ctx))
+def build_metadata(props):
+    metadata = {k : props[k] for k in ["purpose", "maturity"] if k in props}
+    metadata.update(with_cluster(get_cluster_metadata, props))
     return {k : metadata[k] for k in metadata.keys() if metadata[k] is not None}
 
 
-def export_keyspaces(ctx, export_fn, keyspace_objs, metadata, output_dir = None):
-    options = {'rf': ctx.obj['rf']}
-    export_dir = output_dir or ctx.obj["output-dir"]
+def export_keyspaces(props, export_fn, keyspace_map, metadata):
+    options = {'rf': props['rf']}
+    export_dir = props["output-dir"]
 
     if export_dir is None:
-        print(export_fn(keyspace_objs, metadata, options))
+        print(export_fn(keyspace_map.keys(), metadata, options))
     else:
-        for keyspace_obj in keyspace_objs:
-            with open(os.path.join(export_dir, keyspace_obj.name),'w') as output_file:
+        for (keyspace_obj, hashed_ks_name) in keyspace_map.items():
+            # Output filename should use the file-safe alphabet for b64 encoding
+            output_file_name = hash_to_filename(hashed_ks_name) if props["anonymize"] else keyspace_obj.name
+            with open(os.path.join(export_dir, output_file_name),'w') as output_file:
                 log.info("Writing schema for keyspace " + keyspace_obj.name + " to file " + output_file.name)
+                # Metadata string for b64 encoded keyspace name should use the conventional encoding
+                metadata['hashed_ks_name'] = hash_to_str(hashed_ks_name)
```

Comment:
Note that "hashed_ks_name" is added to the metadata in export_keyspaces, meaning it's added every time we do an export regardless of whether it's CQL, Gemini and/or going or not going to Github.  Rationale here is that we want to use the hashed keyspace name as a unique identifier (at least within the domain of "schemas submitted _by this user_").

Further note that "github-user" is added in the actual "contribute" logic itself.  That metadata field has no meaning outside the notion of a Github submission so seemed sane to constrain it there.

This notion of different tiers of metadata is implicit here but probably worthwhile to point out.